lib/Analysis/FIRRTLInstanceInfo.cpp

```diff
@@ -0,0 +1,156 @@
+  bool underDut = false;
+
+  /// Indicates that the current hierarchy is under a layer.
+  bool underLayer = false;
+};
+} // namespace
+
+InstanceInfo::InstanceInfo(Operation *op, mlir::AnalysisManager &am) {
+  auto &iGraph = am.getAnalysis<InstanceGraph>();
+
+  SmallVector<Frame> worklist({{iGraph.getTopLevelNode()}});
+  while (!worklist.empty()) {
+    auto frame = worklist.pop_back_val();
+    auto *node = frame.node;
+    auto moduleOp = node->getModule();
+
+    // Compute information about this instance.
+    bool isDut = AnnotationSet(moduleOp).hasAnnotation(dutAnnoClass);
+
+    // Set the baseline attributes.  This is what the attributes will be if
+    // there is no other, existing attributes.
+    ModuleAttributes attributes({/*isDut=*/isDut});
+    attributes.underDut.merge(frame.underDut);
+    attributes.underLayer.merge(frame.underLayer);
+
+    // Merge the baseline attributes with the existing attributes, if they
+    // exist.
+    auto it = moduleAttributes.find(moduleOp);
+    if (it == moduleAttributes.end())
+      moduleAttributes[moduleOp] = attributes;
+    else {
+      auto &oldAttributes = it->getSecond();
+      oldAttributes.underDut.merge(attributes.underDut);
+      oldAttributes.underLayer.merge(attributes.underLayer);
+      attributes = oldAttributes;
+    }
+
+    for (auto *inst : *frame.node) {
+      auto underDut = frame.underDut || isDut;
+      auto underLayer = frame.underLayer ||
+                        inst->getInstance()->getParentOfType<LayerBlockOp>();
+      worklist.push_back({inst->getTarget(),
+                          /*underDut=*/underDut,
+                          /*underLayer=*/underLayer});
+    }
+  }
+  for (auto *node : llvm::depth_first(iGraph.getTopLevelNode())) {
+    auto moduleOp = node->getModule();
+
+    auto &attributes = moduleAttributes[moduleOp];
+
+    if (AnnotationSet(moduleOp).hasAnnotation(dutAnnoClass)) {
+      attributes.isDut = true;
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "InstanceInfo Analysis Results:\n";
+    DenseSet<Operation *> visited;
+    for (auto *node : llvm::depth_first(iGraph.getTopLevelNode())) {
+      auto moduleOp = node->getModule();
+      if (!visited.insert(moduleOp.getOperation()).second)
+        continue;
+      auto attributes = moduleAttributes[moduleOp];
+      llvm::dbgs() << "  - module: " << moduleOp.getModuleName() << "\n"
+                   << "    isDut:      "
+                   << (attributes.isDut ? "true" : "false") << "\n"
+                   << "    underDut:   " << attributes.underDut << "\n"
+                   << "    underLayer: " << attributes.underLayer << "\n";
+    }
+  });
+}
+
+const InstanceInfo::ModuleAttributes &
+InstanceInfo::getModuleAttributes(FModuleOp op) {
+  return moduleAttributes.find(op)->getSecond();
+}
+
+bool InstanceInfo::isDut(FModuleOp op) { return getModuleAttributes(op).isDut; }
+
+bool InstanceInfo::isUnderDut(FModuleOp op) {
```

Comment:
I'd opt against "maybe/may be" only because it implies the possibility of something being true. I.e., this seems closer to "unknown". Since I couldn't think of something pithy, I opted for more verbose names in 09b03f18a. This uses the verbose convention `atLeastOneInstance<property>` and `allInstances<property>`.

I'm open to alternatives, though.